File: test/PyExtTest/LocalsplusTest.cpp
#include "Catch.hpp"

#include "PythonDumpFile.h"
#include "TestConfigData.h"

#include <globals.h>
#include <PyInterpreterState.h>
#include <PyThreadState.h>
#include <PyFrameObject.h>
#include <PyCodeObject.h>
#include <PyTypeObject.h>
using namespace PyExt::Remote;

#include <utils/ScopeExit.h>
#include <vector>
#include <string>
#include <algorithm>
#include <iterator>
#include <regex>
using namespace std;

TEST_CASE("localsplus_test.py has the expected frames with localsplus.", "[integration][localsplus_test]")
{
	auto dump = PythonDumpFile(TestConfigData::instance().localsplusDumpFileNameOrDefault());

	// Set up pyext.dll so it thinks DbgEng is calling into it.
	PyExt::InitializeGlobalsForTest(dump.pClient.Get());
	auto cleanup = utils::makeScopeExit(PyExt::UninitializeGlobalsForTest);

	auto frames = dump.getMainThreadFrames();
	REQUIRE(frames.size() > 6);

	vector<vector<string>> expectations{
		{
			"f_cellvar",

			R"(localsplus: \{\n)"
			R"(\t'self': <link cmd="!pyobj 0n\d+">&lt;A object&gt;</link>,\n)"
			R"(\t'param1': 'test',\n)"
			R"(\t'param2': \[ 1, 2, 3 \],\n)"
			R"(\t'local1': 5,\n)"
			R"(\t'f_cellfreevar': <link cmd="!pyobj 0n\d+">&lt;function f_cellfreevar&gt;</link>,\n)"
			R"(\t'cell1': <link cmd="!pyobj 0n\d+">&lt;cell object&gt;</link>: \{\n)"
			R"(\t1: 2,\n)"
			R"(\t3: 4,\n)"
			R"(\},\n\})"
		},
		{
			"f_cellfreevar",

			R"(localsplus: \{\n)"
			R"(\t'param': 7,\n)"
			R"(\t'local2': 6,\n)"
			R"(\t'f_freevar': <link cmd="!pyobj 0n\d+">&lt;function f_freevar&gt;</link>,\n)"
			R"(\t'cell2': <link cmd="!pyobj 0n\d+">&lt;cell object&gt;</link>: 9,\n)"
			R"(\t'cell1': <link cmd="!pyobj 0n\d+">&lt;cell object&gt;</link>: \{\n)"
			R"(\t1: 2,\n)"
			R"(\t3: 4,\n)"
			R"(\},\n\})"
		},
		{
			"f_freevar",

			R"(localsplus: \{\n)"
			R"(\t'x': 9,\n)"
			R"(\t'cell2': <link cmd="!pyobj 0n\d+">&lt;cell object&gt;</link>: 9,\n)"
			R"(\})"
		},
	};

	for (auto& exp : expectations) {
		auto& name = exp[0];
		auto& details = exp[1];

		SECTION("Localsplus for frame in method " + name + "().")
		{
			auto frame = std::find_if(begin(frames), end(frames), [&name](auto frame) {
				return frame->code()->name() == name;
				});

			std::regex expectedRegex(details);
			REQUIRE(regex_match((*frame)->details(), expectedRegex));
		}
	}
	
}

File: test/scripts/localsplus_test.py
import win32debug, sys, os


class A(object):
    def __init__(self, a):
        self.a = a

    def f_cellvar(self, param1, param2):
        local1 = 5
        # c is cell variable because it is used in f_cellfreevar
        cell1 = {1: 2, 3: 4}

        def f_cellfreevar(param):
            # cell1 is free variable because it is defined outside f_cellfreevar
            # cell2 is cell variable because it is used in f_freevar
            cell2 = param + cell1[1]
            local2 = 6

            def f_freevar():
                x = cell2
                win32debug.dump_process("localsplus_test.dmp")

            try:
                f_freevar()
            except Exception as e:
                # e is not initialized (null in localsplus) when creating the dump!
                pass

        assert 'cell2' in f_cellfreevar.__code__.co_cellvars
        assert 'cell1' in f_cellfreevar.__code__.co_freevars
        assert f_cellfreevar.__code__.co_nlocals == 4  # param, local2, f_freevar, e
        f_cellfreevar(7)


def main(a):
    a.f_cellvar("test", [1,2,3])


main(A(1))
